src/BotwinExtensions.cs

```diff
@@ -127,17 +128,13 @@ private static void ApplyGlobalBeforeHook(IApplicationBuilder builder, BotwinOpt
         public static void AddBotwin(this IServiceCollection services)
         {
             //Get IAssemblyProvider, if not found register default provider.
+            services.TryAddSingleton<IAssemblyProvider, AssemblyProvider>();
+
             var provider = services.BuildServiceProvider();
```

Comment:
This is pretty gross still...

Reply:
Alternative?

Reply:
Maybe just change this entire pattern. Remove the assembly provider since it's too early to discover services. Make it a `Func` instead of an interface?

Reply:
And resolve the Func here? https://github.com/jchannon/Botwin/blob/master/src/BotwinExtensions.cs#L29

Reply:
Here https://github.com/jchannon/Botwin/blob/master/src/BotwinExtensions.cs#L136.

I'm saying to avoid adding things in the service provider that affect adding more services.

Reply:
Right, so I'm taking that as you wouldn't touch the `IServicesCollection` with `IAssemblyProvider`, you'd have some overloads to `AddBotwin` that take optional `Assemby`s/`Type`s/`IAssemblyProvider`s etc...
Just like AutoMapper does here: https://github.com/AutoMapper/AutoMapper.Extensions.Microsoft.DependencyInjection/blob/master/src/AutoMapper.Extensions.Microsoft.DependencyInjection/ServiceCollectionExtensions.cs#L37

Reply:
I had a play with that approach last night and it seemed to work well. The AddBotwin takes in params Assembly and we use it at that point to find modules etc but I also register IEnumerable<Assembly> so that they can be pulled out in places like this : https://github.com/jchannon/Botwin/blob/master/src/BindExtensions.cs#L25

Reply:
You scan all types in the assembly potentially per request or am I missing something?

Reply:
Ah good point. Maybe best to do that check in the AddBotwin method and register the `IEnumerable< AbstractValidator<>>` and then pull them out on the Bind method

Reply:
I think I will merge this as is then put in the fixes I spiked last night on another commit that removes IAssemblyProvider.  Will cc people on that PR when I do it